{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "## Filtering & Plots\n",
    "\n",
    "This notebook is for streamlining the filtering process that occurs after ipyrad. This way, I can rerun the notebook with different filenames instead of do it from scratch each time.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "**[1]** Change working directory to assembly outfiles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/hgfs/SHARED_FOLDER/Learn_iPyrad/CRAGIG_RUNS_12_py/CG_l1_25_c78_H_outfiles\n"
     ]
    }
   ],
   "source": [
    "cd /mnt/hgfs/SHARED_FOLDER/Learn_iPyrad/CRAGIG_RUNS_12_py/CG_l1_25_c78_H_outfiles/"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CG_l1_25_c78_H.vcf\r\n"
     ]
    }
   ],
   "source": [
    "# get name of VCF file\n",
    "!ls *.vcf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "**[2]** Make directory for new/edited outfiles from ipyrad, called ``new_outfiles``."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "!mkdir new_outfiles"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "**[3]** Use ``vcftools`` to filter for biallelic loci and minor allele frequency. Manual [here](http://vcftools.sourceforge.net/man_latest.html).\n",
    "\n",
    "Be sure to change:\n",
    "- name of input vcf file\n",
    "- name of output vcf file\n",
    "\n",
    "Use: \n",
    "- ``--maf`` for minor allele frequency\n",
    "- ``--recode-INFO-all`` to keep info column data\n",
    "- ``--recode`` to make a new filtered file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "VCFtools - UNKNOWN\n",
      "(C) Adam Auton and Anthony Marcketta 2009\n",
      "\n",
      "Parameters as interpreted:\n",
      "\t--vcf CG_l1_25_c78_H.vcf\n",
      "\t--recode-INFO-all\n",
      "\t--max-alleles 2\n",
      "\t--min-alleles 2\n",
      "\t--out new_outfiles/CG_l1_25_c78_H_biall_NOmaf\n",
      "\t--recode\n",
      "\n",
      "After filtering, kept 25 out of 25 Individuals\n",
      "Outputting VCF file...\n",
      "After filtering, kept 70444 out of a possible 73456 Sites\n",
      "Run Time = 2.00 seconds\n"
     ]
    }
   ],
   "source": [
    "!vcftools --vcf CG_l1_25_c78_H.vcf \\\n",
    "--min-alleles 2 \\\n",
    "--max-alleles 2 \\\n",
    "--out new_outfiles/CG_l1_25_c78_H_biall_NOmaf \\\n",
    "--recode-INFO-all \\\n",
    "--recode"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "**[4]** Get one SNP per RAD tag\n",
    "\n",
    "A collaborator made a script to get one SNP per RAD tag (one with highest coverage). If there's a tie between SNPs, it picks the one closest to the beginning of the read. Script [here](https://github.com/nclowell/RAD_Scallops/blob/master/CRAGIG_combined_lanes/Scripts/subsetSNPs_KS.py).\n",
    "\n",
    "Be sure to change:\n",
    "- name of input VCF file\n",
    "- name of output VCF file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/hgfs/SHARED_FOLDER/Learn_iPyrad/CRAGIG_RUNS_12_py/CG_l1_25_c78_H_outfiles/new_outfiles\n"
     ]
    }
   ],
   "source": [
    "cd new_outfiles"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total SNPS: 70456\r\n",
      "Unlinked SNPs: 7272\r\n"
     ]
    }
   ],
   "source": [
    "!python ../../oneSNP.py \\\n",
    "CG_l1_25_c78_H_biall_NOmaf.recode.vcf \\\n",
    "CG_l1_25_c78_H_biall_NOmaf_oneSNP.vcf"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "**[5]** Make dictionary for translating STR file locus names to VCF file locus names\n",
    "\n",
    "In a few steps, we will use [PGD Spider](http://www.cmpg.unibe.ch/software/PGDSpider/) to convert the filtered VCF file into a STR file, which can be read into ``adegenet`` for population genetic statistics. However, PGD Spider will give loci arbitrary names: SNP_1, SNP_2, SNP_3, etc. So, we will need to give them informative names based on the locus number and position in the VCF file in order to make use of the population genetic statistics.\n",
    "\n",
    "Be sure to change the name of the filtered VCF file. Your filtered vcf file will be the name you gave it plus ``.recode.vcf``"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "vcf_filename = \"CG_l1_25_c78_H_biall_NOmaf_oneSNP.vcf\" # biall, maf, onesnp"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of loci: 7272\n"
     ]
    }
   ],
   "source": [
    "vcf_file = open(vcf_filename,\"r\") # change!\n",
    "vcf_file_lines = vcf_file.readlines()\n",
    "vcf_file.close()\n",
    "\n",
    "# double check number of loci\n",
    "count = 0\n",
    "for line in vcf_file_lines:\n",
    "    if line[0] != \"#\":\n",
    "        count += 1\n",
    "        \n",
    "print \"Number of loci: \" + str(count)\n",
    "\n",
    "locus_names = {}\n",
    "snp_name_count = 1\n",
    "\n",
    "for line in vcf_file_lines:\n",
    "    if line[0] != \"#\":\n",
    "        linelist = line.strip().split()\n",
    "        locus = linelist[0]\n",
    "        pos = linelist[1]\n",
    "        vcf_locus_name = locus + \"_\" + pos\n",
    "        str_locus_name = \"SNP_\" + str(snp_name_count)\n",
    "        snp_name_count += 1\n",
    "        locus_names[str_locus_name] = vcf_locus_name   "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "**[6]** Extract read depth information by locus from VCF file. \n",
    "\n",
    "This block of code will make a dictionary with the VCF locus file names as keys and lists of read depths (across samples) as values. Plot average read depth (across samples) per locus."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAcgAAAEjCAYAAAC2IpGrAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAIABJREFUeJzt3Xm4HFWZx/Hvj7AkQJAlEcMSEhBkwNEIV9ARkJFlWBxZ\nZIuKgAyBEQZ1dIZFhsUBBEdwZFAQJLIHUEBQcWQT0VGWABHCJluAhEAiawJhSXjnj3OaVJrqu+Xe\nru7O7/M897m119vVVfXWOXWqWhGBmZmZLWqpqgMwMzNrRU6QZmZmJZwgzczMSjhBmpmZlXCCNDMz\nK+EEaWZmVmKJTZCSbpH0T1XH0QySzpd0Yi+nHSMpJC09SLGEpPf3clpJ+omkFyXdMRjxVGlJ2gcH\nS3fbUNJoSXMlDenFcraWNL3QP03StgMZa7NI+oSkR/Jn37XqeFqRpOUl/UrSy5ImNZqu1wky74gv\nSlpuYEK0TjcACWALYDtgrYjYbIDC6hNJoySdJ2mmpDmSHpJ0gqQV8nhJOkzSvZJek/Rs/tz79GM9\n10p6Jl9EjOnlfN/NJ8NabF+sGx+SXs0ny7mSftyXuOqWtWVhObW/kPTZPF6STpQ0I594bpG0cS+W\n+679pD5h9UdEPBURK0bEgsVZTiOSjs+ff6/CsKXrvz9Jfyfp5vwdvSzpF5I2GoyYsm8BZ+bP/vNB\nXE8pSStLmpiPhTmS/iLpyML4kHSfpKUKw06UdH7url2k1/axacX5B8jewKrAahExvtFEvUqQ+cve\nEgjgMwMQXNk6BqXE0m7ySWaJLdnXWQeYFhGvlo0c7H1G0qrAn4BhwMcjYjgpYa8MrJcnOwP4KvB1\nYDVgTeAYYIc+ru5t4H+Bz/ZxvleBfwTeA+wHfF/S39VN8+F8slwxIvp9wRIRvy8sZ0Xg08DcHDfA\nnsCXSOeK2ra7qL/raxMvACc0KqVK+jhwPXANsAYwFvgz8H+S1h2kmNYB7h+kZffG94AVgb8h7Zef\nAR6tm2YNoKeLyJXzfjYeOFZSX4+p7qwDPBwR87udKiJ6/AOOBf4POB34ZWH45sCzwJDCsN2Ae3P3\nUsCRwGPA88AVwKp53BhSwj0QeAq4NQ//aV7my8CtwMaFZa8G/AJ4BbgTOBH4Q2H8hsANpJ32YWCv\nbj7TLcA/FeI8BngSmAVcCLynMO0WwB+Bl4Cngf3rl5H796/FA4i0o8zK8d4HfLCbWE7K23ge8H7S\njnUeMBOYkT/rkDz9esDNeZv+FbiEtDPVlvcR4G5gDnA5cBlwYoN1DwG+m5fzOHBo/l6WzuO7i2P/\nHPOZ+ft6CNgmjzsJWAC8TjqJnpmHB3AI8Ejenj8AVBLXgXneBXn+E4CtgenAEXkfuShPexDpAHwB\nuBZYo7CcAL6c1zcH+M+8/f6Yv5crgGUbbJsT8/e2VIPxG+T4unpzHHW3DxaGLZ1jHtPPZV4LfL3u\n87+/j8sYBpxGOh5eBv4ADCuZ7ifATwr9RwBXFPo3Bl7v53bYGpjey3m/DdyRv89rePc5prYvHwA8\nmPeDx4GDG60PmAZs28O6jycde38G9iv7/oDfAz8smffXwIXFdQP/TjpfzAR2BXYC/pL366ML825G\nuvh4KU97Zm0fJp1r3yadR+YCy3UT/xp5f3mBdPwcVPfZriCdC+eQEm5X3bxXArOBJ4DDC+OmArt2\ns97I+8ojhe/mROD8su8tD7sT+EbJsh4Bdij0L5s/z4dy/1bAbaT9+GlgX9K56U3grbyN9msYay8P\nmEdJJ5lN80JXL4x7DNiu0P9T4Mjc/ZUc3FrAcsCPgEl1G+FCYAXyAUi6Ah2ep/9vYEph2Zflv+WB\njfIHriWkFXL/AaSd9COkk/5GPR2UeZ2PAuuSrnyuYuHJd528g4wHliEl6XFlBzaLJsh/AO4ilTZE\nupoa1U0sT5FOKEvn9Vydt9cKwHtJJ4CD8/TvJ5VklgNGki4k/ruwgzwJfC0vZ4/8nTVKkIeQEtva\npKv+37LoSaW7OPYH5hfWtTdpR1y1bPsUDo5f5u0ymnSA7dAgtne2Z+FEMh84NX/2YcCn8ve8SR72\nP+SLrcL6rgFWytv3DeCm/F2/B3iABgcIad89oZvj4hBSCbfPiayb7dPvBJm3x0wWPWEE8AzpguKq\n3iyXdNFyC6k0PAT4O+pOtHl/mANsXRi2Dmmf3yDvD98Bft7P7bA1vU+QM4AP5piuBC6uO8fU9uWd\nSRdHAj4JvAZsUrY+ep8gLyaVkB7Pn/md7490nloA/H3JvAcAM+v262PzMg4iHReXks6FG5MS3tg8\n/abAx/K6xpCS/lf7Enue7lbgh8BQYFxe56cKn+11UpIeQroIuS2PWyp/z8eSzjfr5s//D3n8j0kJ\n9QBg/ZL1BrB+XkbtHFyaIPN39Yn8XW1TsqxvARcU+ncBpubusaQEuFde1ggWnrvfWV+326gXG3EL\n0gl2RO5/CPhaYfyJwMTcPZxU5bNO7n+w+KGAUXlZtS82gHW7WffKeZr35C/pLeADdeuuJaS9gd/X\nzf8j4LieDkrSCfPLhXEfKMR5FHB1bw5sFk2QnyJd/X2MBiWQuuV8q9C/OulEPqwwbDzw2wbz7wrc\nk7u3Ip0QVRj/RxonyJuBQwr92xd2zm7jyJ+3fl13APuWbZ/CwbFFof8K8gVVSWzvbM/CieRNYGhh\n2HnAdwr9K+bvbkxhfZ8ojL8LOKLQfxr54qJk/Y8Ut03J+GPIJ43CsOmkK/vXycdBD9/7QCbIC0jV\nncXvYyvSSWxlUkljKoUr85JlLEU6GX+4h3XtSyo5FNe1LPD9HP/8PH5sL+K+hXQCfKnwN5feJ8hT\nCv0b5X1kCCUlkbp5fw58pbBv9StB5u7bgX9m0QS5Vu7esGTeHYC3Cuuex8KameF5vs3r9tvSUhmp\niv/qPsa+Nil5Dy8M+zYLk9TxwI1123Ve7t4ceKpueUeRaxNIF2pH55jfIhU+dixMG6SL/J1IF/PL\nUp4gXwJeJOWRwxt8jg1JF+VDc//l5NI28B/ATxvM16sE2Zt7XfsB10fEX3P/pXkYhf7dc+Od3YG7\nI+LJPG4d4GpJL0l6KX/QBaQTb83TtQ5JQySdIukxSa+QvmhImX8kaed7umzevK7Na+vK6/s88L5e\nfMY1SF9UzZMsTBBrk0rJfRIRN5NOSD8AZkk6R9JK3cxS/1mWAWYWPsuPSCU4JK0u6bLcGOIV0lXs\niMJnmRF5Lyh8nkbWqFt3cdpu48jK1rVGN+uDVJqpeY2U1HprdkS8Xuhf5LuLiLmkquc1C9M8V+ie\nV9LfaP3Pky7qGnnX+IhYi/RdLEe6+m0KSf9FKkXtVfw+IuLWiHgzIl4i1eiMJdVmNDKCVKLoaZ/f\nj1RFWPzujwU+SjpmhpKqxW+WtHwvPsLhEbFy7Y90f7O36vffZVh4PLxD0o6SbpP0Qt6Xdyqbrp+O\nAb5J+tw1L5KqO8v2oVGkmo+a52NhY6J5+X/pfippA0m/zI1gXgFOpu+fYw3ghYiYUxj2JIseN/XH\n6dB8338dYI26c+3R5PN6RMyLiJMjYlNSjdsVwE/zPf13RMR1pAvKgxvEOCIiVomIv4mIM8omiIiH\nSPvqzpJq98UvzaP7de4u6jZBShpGKp5+Mn8Zz5Kq0z4s6cM5wAdIG3ZH4HOF4CDtuDsWd/yIGBoR\nM4qfsdD9OVIReVtSqXFMLRRS8X8+6aqsZu26df2ubl0rRsQ/97wZeIb0pdeMzut6Li93vbKZSKXl\n4sG/SDKOiDPyTrIRqdrp37qJobgdniaV3EYUPstKEVFrEXhynv5vI2Il4AssPBnPBNaUVDw5j+5m\nvTNZdDsWp+0pDhqs65mSzzRQ6pe5yHen1Lp0NVK12+K6Editm0ZTNwNrSeoagHX1m6QTSMff9hHx\nSg+TB90n7r+SSr+N9nkkrU0q9VxYN2occHlETI+I+RFxPrAKaf8fTPX771ssmnzIF/BXku63r56T\n8HUM0EVMRNzAwltRtWGvku4V7lkyy16kmqv+OItUk7d+Pv6Ppu+f4xlgVUnDC8NG07vj5mngibpz\n7fCI2Kl+wrw/nkyq/h5bsqxv5vh7cxHVyCRSzdZupFty0wpxNtyPe6OnEuSupBLfRqSdfxzp6vP3\nQLE5+aWkq9OtSPcga84GTpK0DoCkkZJ26WZ9w0kn5OdJG+zk2oh8dXUVcLzSMywb1sXwS2ADSftK\nWib/fVRSd1fLNZOAr0kam69CTiYd6PNJN+G3lbRXbsK9mqRxeb4ppNLz8krP9h1YW2Be9+aSliEl\n0tdJV5M9ioiZpJZvp0laSdJSktaT9MnCdpoLvCxpTRZNvH8iJffD8zbYnXRTv5Er8rRrSVqF1Kiq\nt3FAKk3W1rUnaf+4Lo97jnR/YjBNAg6QNC6fBE8Gbi8cJIvjdNK9ywsK+/Cakk6X9KGIeJhUor5M\n0naShuXWjPWtSHtF0lBSyRNgudzf0zxHkS4st42I5+vGbZy3y5C8X59GOgE+2Gh5EfE2MBE4XdIa\ned6Pa9HHu/YF/hgR9VfndwJ75hqOpSTtSyrN1bdgHGhfkLRRLql+C/hZvPvRjmVJ23Y2MF/SjqTb\nCQPpm6SGNkVHAvtJOlzScEmrKD2T/HFSCbs/hpMaJM3N58HeFAIWERFPk269fFvSUEkfIp2/Lu7F\n7HcAcyQdUdvnJX1Q0kcBJP1HPv8tm/fhr5CqSx8uieMWUrX/fvXj+mAS6QJxAosW0C4GdpD02Xzu\nHlEr2PVWTwlyP1K98lMR8Wztj1R1+HktbGY/iXTT++ZCVSyk+xHXAtdLmkNq9LB5N+u7kFQanUFq\nPHFb3fjDSCXLZ0nNxyeREiq5qmB7UtPhWqOEWmOOnkzMy7uVdN/kdeBf8nKfIlXFfJ3UOmoKUNvI\n3yPd73iOdP/nksIyVwLOJVWzPElK+v/Vi1hqvkg6qB/Iy/gZC6tqTiA1SnkZ+BXpwoEc75ukqu79\nc7x7F8eXOBf4Dakl3t0l03YXB6R7L+uTrthPAvYonKi/D+yh9PxsaRXJ4oqIG0n3Gq4klYbXo+fm\n471d9gukZPcWcHveh28ibffaSf9Q0qMep5O293RSS9m9SQ2v+qLW8hBSCWFeN9PWnEy68n9UC58b\nOzqPW510T+YVUiOKMcCnI+KtHpb5DVLr3TtJn+lUFj1XfJG0v9c7lbQfTSGdEL8GfDZX7w6mi4Dz\nScf8UODw+gny+eFw0gXhi6SLimsHMoiI+D9S8igO+wOpwd7upP3zSVIDwi0i4pF+ruobpPjnkI7f\ny/u5nPGkfeIZUmO84/Lx1K188fFpUoHpCdKx/2PSuRlSLcVP8vBnSA0Kd863P8ocQ2og2CuSHpa0\ndyGe6cBkUnuPKwrDnyA9AnUEaT++G/jb3q4H8g32diXpVOB9EbE4Vx/WT5L2JzUy2aLqWMzMBlpb\nPZAuaUNJH1KyGalK4Oqq4zIzs87TVgmSVPd+Feme3uWkeyrXVBqRWTf07lez1f62bOYySpZ5f4Nl\nfr6/y+xhff3+DIPx+fsQ968brPvonueuXpXbrhO0dRWrmZnZYGm3EqSZmVlTOEGamZmVcII0MzMr\n4QRpZmZWwgnSzMyshBOkmZlZCSdIMzOzEk6QZmZmJZwgzczMSjhBmpmZlXCCNDMzK+EEaWZmVsIJ\n0szMrIQTpJmZWYmlqw6gVYwYMSLGjBlTdRhmZm3lrrvu+mtEjKw6jsHgBJmNGTOGyZMnVx2GmVlb\nkfRk1TEMFlexmpmZlXCCNDMzK+EEaWZmVsIJ0szMrIQTpJmZWQknSDMzsxJOkGZmZiWcIM3MzEo4\nQZqZmZXwm3TMOtiYI3+1SP+0U3auKBKz9uMSpJmZWQknSDMzsxJOkGZmZiWcIM3MzEo4QZqZmZVw\nK1azJUixVatbtJp1zyVIMzOzEm2RICVNlDRL0tTCsMslTcl/0yRNycPHSJpXGHd2dZGbmVm7apcq\n1vOBM4ELawMiYu9at6TTgJcL0z8WEeOaFp2ZmXWctkiQEXGrpDFl4yQJ2Av4VDNjMjOzztYWCbIH\nWwLPRcQjhWFjJd0DvAIcExG/L5tR0gRgAsDo0aMHPVCzVuLX0Jl1ry3uQfZgPDCp0D8TGB0RHwH+\nFbhU0kplM0bEORHRFRFdI0eObEKoZmbWLto6QUpaGtgduLw2LCLeiIjnc/ddwGPABtVEaGZm7aqt\nEySwLfBQREyvDZA0UtKQ3L0usD7weEXxmZlZm2qLBClpEvAn4AOSpks6MI/ah0WrVwG2Au7Nj338\nDDgkIl5oXrRmZtYJ2qKRTkSMbzB8/5JhVwJXDnZMZmbW2doiQZpZY4Px+ji3cDVrkypWMzOzZnOC\nNDMzK+EEaWZmVsIJ0szMrIQTpJmZWQknSDMzsxJOkGZmZiWcIM3MzEo4QZqZmZVwgjQzMyvhBGlm\nZlbC72I16yD171A1s/5zgjRrA4PxQvJ2Wr9ZFVzFamZmVsIJ0szMrIQTpJmZWQknSDMzsxJOkGZm\nZiXaphWrpInAp4FZEfHBPOx44CBgdp7s6Ii4Lo87CjgQWAAcHhG/aXrQZm3Ej4iYLaqdSpDnAzuU\nDP9eRIzLf7XkuBGwD7BxnueHkoY0LVIzM2t7bZMgI+JW4IVeTr4LcFlEvBERTwCPApsNWnBmZtZx\n2iZBduMwSfdKmihplTxsTeDpwjTT87BFSJogabKkybNnz64fbWZmS7B2T5BnAesB44CZwGl9mTki\nzomIrojoGjly5GDE1/LGHPmrd/7MzGyhtk6QEfFcRCyIiLeBc1lYjToDWLsw6Vp5mJmZWa+0dYKU\nNKrQuxswNXdfC+wjaTlJY4H1gTuaHZ+ZmbWvdnrMYxKwNTBC0nTgOGBrSeOAAKYBBwNExP2SrgAe\nAOYDh0bEgiribjV9qUr1C6rNbEnWNgkyIsaXDD6vm+lPAk4avIjMquH7xWbN0dZVrGZmZoPFCdLM\nzKxE21Sxmi1JXI1qVj2XIM3MzEo4QZqZmZVwgjQzMyvhe5BmA8zPj5p1BpcgzczMSrgEae9wy0nr\njfr9xKVk61QuQZqZmZVwgjQzMyvhKlazJuquAY+ruM1ai0uQZmZmJZwgzczMSriK1WwQudrUrH05\nQVqvuGm/NeIXI1inchWrmZlZCSdIMzOzEk6QZmZmJdoiQUqaKGmWpKmFYf8l6SFJ90q6WtLKefgY\nSfMkTcl/Z1cXuZmZtat2aaRzPnAmcGFh2A3AURExX9KpwFHAEXncYxExrrkhmpkbc1knaYsEGRG3\nShpTN+z6Qu9twB7NjMmWbG652TveTtbO2qKKtRe+BPy60D9W0j2Sfidpy0YzSZogabKkybNnzx78\nKM3MrG20fYKU9E1gPnBJHjQTGB0RHwH+FbhU0kpl80bEORHRFRFdI0eObE7AZmbWFtqiirURSfsD\nnwa2iYgAiIg3gDdy912SHgM2ACZXFWfVOuFtLr63ZWbN1rYJUtIOwL8Dn4yI1wrDRwIvRMQCSesC\n6wOPVxSmWUOdcOFi1snaIkFKmgRsDYyQNB04jtRqdTngBkkAt0XEIcBWwLckvQW8DRwSES9UEriZ\nmbWttkiQETG+ZPB5Daa9ErhycCMyM7NO1/aNdMzMzAZD00uQkv42Iu5r9nqtNfk5OTNrVVVUsf5Q\n0nKkt+NcEhEvVxCDmTWZWyJbu2l6goyILSWtT3q4/y5JdwA/iYgbmh2LLdl8wjaz7lRyDzIiHgGO\nIb079ZPAGfnF47tXEY+ZmVm9Ku5Bfgg4ANiZ9MLxf4yIuyWtAfwJuKrZMVnf+d6hmXW6Ku5B/g/w\nY+DoiJhXGxgRz0g6poJ4bDEt6VWVfuDfrDNVkSB3BuZFxAIASUsBQyPitYi4qIJ4zMzM3qWKe5A3\nAsMK/cvnYWZmZi2jigQ5NCLm1npy9/IVxGFmZtZQFQnyVUmb1HokbQrM62Z6MzOzpqviHuRXgZ9K\negYQ8D5g7wri6FhuNGJmtviqeFHAnZI2BD6QBz0cEW81Ow5rPUt6a1gzay1V/ZrHR4Exef2bSCIi\nLqwoFmtRLgmbWZWqeFHARcB6wBRgQR4cgBOkmZm1jCpKkF3ARhERFazblnC9LZW6utfMqmjFOpXU\nMMfMzKxlVVGCHAE8kH/F443awIj4TAWx2CBzSczM2lUVCfL4CtZpZmbWJ02vYo2I3wHTgGVy953A\n3T3NJ2mipFmSphaGrSrpBkmP5P+r5OGSdIakRyXdW3wxgZmZWW9U0Yr1IGACsCqpNeuawNnANj3M\nej5wJou2dj0SuCkiTpF0ZO4/AtgRWD//bQ6clf+b9YsfOTFb8lRRxXoosBlwO6QfT5b03p5miohb\nJY2pG7wLsHXuvgC4hZQgdwEuzC1lb5O0sqRRETFzID6AmS0+/6aotboqEuQbEfGmJAAkLU16DrI/\nVi8kvWeB1XP3msDThemm52GLJEhJE0ilWUaPHt3PEKxe1aUtn3jNbCBUkSB/J+loYJik7YAvA79Y\n3IVGREjqU6KNiHOAcwC6urr8XGYHqjpZm1n7quI5yCOB2cB9wMHAdcAx/VzWc5JGAeT/s/LwGcDa\nhenWysPMzMx6pYpWrG9HxLkRsWdE7JG7+1t6uxbYL3fvB1xTGP7F3Jr1Y8DLvv9oZmZ9UUUr1ico\nuecYEev2MN8kUoOcEZKmA8cBpwBXSDoQeBLYK09+HbAT8CjwGnDAQMVvZmZLhqrexVozFNiT9MhH\ntyJifINR73o8JJdID+1XdGZmZlRTxfp84W9GRPw34KaGZmbWUqqoYi2+1WYpUomyqt+lNDMzK1VF\nYjqt0D2f9Nq5vconNbMlQU+P4/h5VqtC0xNkRPx9s9dpnc3POprZYKiiivVfuxsfEac3KxazIida\nMyuqqhXrR0nPKgL8I3AH8EgFsZiZmZWqIkGuBWwSEXMAJB0P/CoivlBBLGZmZqWqeNXc6sCbhf43\nWfiScTMzs5ZQRQnyQuAOSVfn/l1JP1VlZmbWMqpoxXqSpF8DW+ZBB0TEPc2Ow8zMrDtVVLECLA+8\nEhHfB6ZLGltRHGZmZqWaniAlHQccARyVBy0DXNzsOMzMzLpTRQlyN+AzwKsAEfEMMLyCOMzMzBqq\nIkG+mX9tIwAkrVBBDGZmZt2qIkFeIelHwMqSDgJuBM6tIA4zM7OGlApzTV6ptB2wPSDgNxFxQ9OD\nqNPV1RWTJ0+uOowB4Vem2ZLELzKvlqS7IqKr5ynbT1Mf85A0BLgxv7C88qRoZmbWSFOrWCNiAfC2\npPc0c71mZmZ9VcWbdOYC90m6gdySFSAiDu/PwiR9ALi8MGhd4FhgZeAgYHYefnREXNeviM3MbIlT\nRYK8Kv8NiIh4GBgH71ThzgCuBg4AvhcR3x2odZmZ2ZKjaQlS0uiIeCoiBvO9q9sAj0XEk5IGcTVm\nZtbpmnkP8ue1DklXDtI69gEmFfoPk3SvpImSVhmkdZqZWQdqZhVrsUi37oAvXFqW9Iae2ivszgL+\nk/RCgv8ETgO+VDfPBGACwOjRowc6pKbyox1mZgOrmSXIaNA9UHYE7o6I5wAi4rmIWBARb5NeRLDZ\nuwKKOCciuiKia+TIkYMQkpmZtatmliA/LOkVUklyWO4m90dErLSYyx9PoXpV0qiImJl7dwOmLuby\nzcxsCdK0BBkRQwZr2fl9rtsBBxcGf0fSOFJpdVrdODPrEMXbC36rjg2kKh7zGHAR8SqwWt2wfSsK\nx8zMOkBVP5hsZmbW0pwgzczMSjhBmpmZlXCCNDMzK+EEaWZmVsIJ0szMrIQTpJmZWQknSDMzsxJO\nkGZmZiWcIM3MzEo4QZqZmZVwgjQzMyvRES8rXxL5B5LNzAaXS5BmZmYlnCDNzMxKOEGamZmVcII0\nMzMr4QRpZmZWwgnSzMysREc85iFpGjAHWADMj4guSasClwNjgGnAXhHxYlUxmplZe+mkEuTfR8S4\niOjK/UcCN0XE+sBNud/MzKxXOilB1tsFuCB3XwDsWmEsZmbWZjolQQZwvaS7JE3Iw1aPiJm5+1lg\n9WpCMzOzdtQR9yCBLSJihqT3AjdIeqg4MiJCUtTPlJPpBIDRo0c3J1IzM2sLHVGCjIgZ+f8s4Gpg\nM+A5SaMA8v9ZJfOdExFdEdE1cuTIZoZsZmYtru1LkJJWAJaKiDm5e3vgW8C1wH7AKfn/NdVFaWZV\nKL7Uf9opO1cYibWjtk+QpHuLV0uC9HkujYj/lXQncIWkA4Engb0qjNHMzNpM2yfIiHgc+HDJ8OeB\nbZofkZmZdYK2T5BmZr1R/xuqrnK1nnREIx0zM7OB5hJkm6i/+jWzd/NxYgPJJUgzM7MSTpBmZmYl\nnCDNzMxKOEGamZmVcII0MzMr4QRpZmZWwgnSzMyshBOkmZlZCSdIMzOzEk6QZmZmJZwgzczMSvhd\nrGZmdfxDywYuQZqZmZVyCdLMlnj+FRAr4wRpZkskJ0XriatYzczMSrR1CVLS2sCFwOpAAOdExPcl\nHQ8cBMzOkx4dEddVE2X/+QrXzKw6bZ0ggfnA1yPibknDgbsk3ZDHfS8ivlthbGZm1sbaOkFGxExg\nZu6eI+lBYM1qozIzs07QMfcgJY0BPgLcngcdJuleSRMlrVJZYGZm1pY6IkFKWhG4EvhqRLwCnAWs\nB4wjlTBPazDfBEmTJU2ePXt22SRmZraEavsEKWkZUnK8JCKuAoiI5yJiQUS8DZwLbFY2b0ScExFd\nEdE1cuTI5gVtZmYtr60TpCQB5wEPRsTpheGjCpPtBkxtdmxmZtbe2rqRDvAJYF/gPklT8rCjgfGS\nxpEe/ZgGHFxNeGbW7uoft/K7WZccbZ0gI+IPgEpGtd0zj2Zm1lrauorVzMxssDhBmpmZlXCCNDMz\nK+EEaWZmVsIJ0szMrERbt2JtF8Vm4m4ibtbefDwvOZwgm6y7Z6r881Zm7cXPSHY2V7GamZmVcII0\nMzMr4SrWirla1cysNbkEaWZmVsIJ0szMrISrWM3MBkhfWqm7xWvrcwnSzMyshBOkmZlZCVexmplV\nwG/kaX1OkGZmg8SPcbU3J0gzsxbj0mVrcII0M6tYX0qaTp7N09EJUtIOwPeBIcCPI+KUwVqXq1LM\nzDpLx7YEb9DHAAAH10lEQVRilTQE+AGwI7ARMF7SRtVGZWZm7aJjEySwGfBoRDweEW8ClwG7VByT\nmZm1iU6uYl0TeLrQPx3YvKJYzMz6pbvbN315O4/vXfZdJyfIHkmaAEzIvXMlPbyYixwB/HUxl9Fs\n7Raz4x1c7RYvtF/MgxavTh3Y6bKe4l2nT0trI52cIGcAaxf618rD3hER5wDnDNQKJU2OiK6BWl4z\ntFvMjndwtVu80H4xO9720cn3IO8E1pc0VtKywD7AtRXHZGZmbaJjS5ARMV/SYcBvSI95TIyI+ysO\ny8zM2kTHJkiAiLgOuK6Jqxyw6tomareYHe/gard4of1idrxtQhFRdQxmZmYtp5PvQZqZmfWbE+QA\nkbSDpIclPSrpyKrjKSNpmqT7JE2RNDkPW1XSDZIeyf9XqTjGiZJmSZpaGFYao5Iz8ja/V9ImLRLv\n8ZJm5O08RdJOhXFH5XgflvQPFcS7tqTfSnpA0v2SvpKHt+Q27ibeltzGkoZKukPSn3O8J+ThYyXd\nnuO6PDccRNJyuf/RPH5Mi8R7vqQnCtt3XB5e+THXVBHhv8X8IzUCegxYF1gW+DOwUdVxlcQ5DRhR\nN+w7wJG5+0jg1Ipj3ArYBJjaU4zATsCvAQEfA25vkXiPB75RMu1Ged9YDhib95khTY53FLBJ7h4O\n/CXH1ZLbuJt4W3Ib5+20Yu5eBrg9b7crgH3y8LOBf87dXwbOzt37AJc3efs2ivd8YI+S6Ss/5pr5\n5xLkwGjn19rtAlyQuy8Adq0wFiLiVuCFusGNYtwFuDCS24CVJY1qTqRJg3gb2QW4LCLeiIgngEdJ\n+07TRMTMiLg7d88BHiS9daolt3E38TZS6TbO22lu7l0m/wXwKeBneXj99q1t958B20hSk8LtLt5G\nKj/mmskJcmCUvdauu4O4KgFcL+mu/BYhgNUjYmbufhZYvZrQutUoxlbe7oflKqiJhWrrloo3V+d9\nhFRqaPltXBcvtOg2ljRE0hRgFnADqRT7UkTML4npnXjz+JeB1aqMNyJq2/ekvH2/J2m5+nizVjrm\nBpwT5JJli4jYhPQLJ4dK2qo4MlIdSks3a26HGIGzgPWAccBM4LRqw3k3SSsCVwJfjYhXiuNacRuX\nxNuy2zgiFkTEONLbuzYDNqw4pG7Vxyvpg8BRpLg/CqwKHFFhiJVxghwYPb7WrhVExIz8fxZwNeng\nfa5WRZL/z6ouwoYaxdiS2z0inssnnbeBc1lYxdcS8UpahpRsLomIq/Lglt3GZfG2+jYGiIiXgN8C\nHydVRdaeOy/G9E68efx7gOebHCqwSLw75KrtiIg3gJ/Qgtu3GZwgB0bLv9ZO0gqShte6ge2BqaQ4\n98uT7QdcU02E3WoU47XAF3PLuo8BLxeqCStTd09mN9J2hhTvPrnl4lhgfeCOJscm4DzgwYg4vTCq\nJbdxo3hbdRtLGilp5dw9DNiOdN/0t8AeebL67Vvb7nsAN+cSfJXxPlS4WBLpfmlx+7bcMTdoqm4l\n1Cl/pNZdfyHdb/hm1fGUxLcuqXXfn4H7azGS7nfcBDwC3AisWnGck0hVZm+R7m8c2ChGUku6H+Rt\nfh/Q1SLxXpTjuZd0QhlVmP6bOd6HgR0riHcLUvXpvcCU/LdTq27jbuJtyW0MfAi4J8c1FTg2D1+X\nlKgfBX4KLJeHD839j+bx67ZIvDfn7TsVuJiFLV0rP+aa+ec36ZiZmZVwFauZmVkJJ0gzM7MSTpBm\nZmYlnCDNzMxKOEGamZmVcII06wdJC/KvHEyV9Ivas2QDsNz9JZ3Zi+mmSRqxGOtYYyCWZdbJnCDN\n+mdeRIyLiA+SXlZ+aNUB9cH+wBo9TWS2pHOCNFt8f6LwwmZJ/ybpzvyi5xMKw3+eXxR/f+Fl8Ug6\nQNJfJN0BfKJsBZJWk3R9nvfHpAe2a+O+kH/Tb4qkH0kakofPzS+avl/STfmtKXsAXcAlefpheTH/\nIulupd8Lbel3h5o1ixOk2WLIyWgb8qsFJW1Per3ZZqQXaW9aeCn8lyJiU1KCOjwnvVHACaTEuAXp\n9wzLHAf8ISI2Jr1Hd3Re398AewOfiPTC6QXA5/M8KwCT8zy/A46LiJ8Bk4HP5xLwvDztXyO9yP4s\n4BuLu13MOsHSPU9iZiWG5Z8IWpP0rs0b8vDt8989uX9FUsK8lZQUd8vD187D3wfcEhGzASRdDmxQ\nsr6tgN0BIuJXkl7Mw7cBNgXuzD8jOIyFLxp/G7g8d18MXEVjtXF31dZjtqRzgjTrn3kRMU7S8sBv\nSPcgzyBVfX47In5UnFjS1sC2wMcj4jVJt5Dew7m4BFwQEUf1Ytru3iv5Rv6/AJ8XzABXsZotloh4\nDTgc+Hr+uaLfAF/Kv1+IpDUlvZf0M0Yv5uS4IfCxvIjbgU/m6tZlgD0brOpW4HN5mTsCtR8IvgnY\nI68DSatKWiePW4qFvyDxOeAPuXsOMHwxP7pZx/OVotliioh7JN0LjI+Ii/J9wT/lKs+5wBeA/wUO\nkfQg6VcmbsvzzpR0PKmhz0ukX6socwIwSdL9wB+Bp/L8D0g6Brhe0lKkXxU5FHgSeJX0A7jHkKpd\n987LOh84W9I80m8VmlkJ/5qHWYeSNDciVqw6DrN25SpWMzOzEi5BmpmZlXAJ0szMrIQTpJmZWQkn\nSDMzsxJOkGZmZiWcIM3MzEo4QZqZmZX4f+QXRyj/2LtAAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7f7465932b90>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sample_list = []\n",
    "snp_name_list = []\n",
    "rdd = {} \n",
    "\n",
    "for line in vcf_file_lines:\n",
    "    if line[0:1] == \"#C\":\n",
    "        linelist = line.strip().split()\n",
    "        sample_list += linelist[9:]\n",
    "    elif line[0] != \"#\": # ignore header lines\n",
    "        genblocks = line.strip().split()\n",
    "        snp_name = genblocks[0] + \"_\" + genblocks[1] # name format e.g., locus_47_1\n",
    "        snp_name_list.append(snp_name)\n",
    "        for genblock in genblocks[9:]: # start on genotype lines\n",
    "            genblocklist = genblock.split(\":\")\n",
    "            rd = int(genblocklist[1])\n",
    "            if snp_name not in rdd:\n",
    "                rdd[snp_name] = [rd]\n",
    "            elif snp_name in rdd:\n",
    "                rdd[snp_name] += [rd]\n",
    "            else:\n",
    "                print \"something funky going on\"\n",
    "                \n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "\n",
    "locus_rd_avs = []\n",
    "for snp in snp_name_list:\n",
    "    locus_rd_avs.append(np.mean(rdd[snp]))\n",
    "    \n",
    "plt.hist(locus_rd_avs, bins = np.arange(0,max(locus_rd_avs),4)-2)\n",
    "plt.suptitle(\"Average locus read depth from \" + vcf_filename)\n",
    "plt.xlabel(\"Read depth\")\n",
    "plt.ylabel(\"Frequency\")\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "**[7]** Make STR file out of filtered VCF using PGD Spider. \n",
    "\n",
    "This is done in the GUI.\n",
    "\n",
    "Select input filetype as VCF, and output filetype as STR. Select your filtered VCF file as your input file, and select a name for your output file, including the .str extension. For the VCF Parser Questions, default parameters are fine. For the STR Writer Questions, be sure to change data type to SNPs.\n",
    "\n",
    "Enter your output structure file name here:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "str_filename = \"CG_l1_25_c78_H_biall_NOmaf_oneSNP.str\" # change!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "![img](https://github.com/nclowell/RAD_Scallops/blob/master/CRAGIG_combined_lanes/Notebooks/imgs_for_notebooks/vcf_parser.png?raw=true)\n",
    "\n",
    "\n",
    "![img](https://github.com/nclowell/RAD_Scallops/blob/master/CRAGIG_combined_lanes/Notebooks/imgs_for_notebooks/snp_writer.png?raw=true)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "**[8]** Rename header of STR file using dictionary created in Step 4.\n",
    "\n",
    "Enter the name for your STR filename with new locus names here."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "new_str_filename = \"CG_l1_25_c78_H_biall_NOmaf_oneSNP_inames.str\" # change!"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "str_file = open(str_filename, \"r\")\n",
    "str_file_lines = str_file.readlines()\n",
    "str_file.close()\n",
    "\n",
    "new_str_header = \"\\t\\t\" # will need to change this line when I get a hold of better code for one SNP/tag\n",
    "old_str_header_list = str_file_lines[0].strip().split()\n",
    "for snpname in snp_name_list:\n",
    "    new_str_header += snpname + \" \"\n",
    "new_str_header = new_str_header[:-1]\n",
    "new_str = open(new_str_filename, \"w\")\n",
    "new_str.write(new_str_header + \"\\n\")\n",
    "for line in str_file_lines[1:]:\n",
    "    new_str.write(line)\n",
    "new_str.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "**[9]** Add population information (OPTIONAL)\n",
    "\n",
    "When I have all of my samples, I'll definitely want to include population information in the STR file. However, as I have small sample sizes per population now, I'd rather leave it out and check out my data as if all populations are one.\n",
    "\n",
    "Be sure to change the name of your population map and name of STR file with population info here:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "popmap_name = \"popmap_cragig_sansAs.txt\"\n",
    "str_filename_w_pops = \"cragig_lanes_12_002_biall_maf_onesnp__inames_wpops.str\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 156,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "popfile = open(\"../\" + popmap_name,\"r\")\n",
    "popfile_lines = popfile.readlines()\n",
    "popfile.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 157,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "str_onesnp_file = open(new_str_filename, \"r\")\n",
    "str_onesnp_file_lines = str_onesnp_file.readlines()\n",
    "str_onesnp_file.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 158,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "pop_dict = {}\n",
    "sample_list = []\n",
    "\n",
    "for line in popfile_lines:\n",
    "    linelist = line.strip().split()\n",
    "    sample = linelist[0]\n",
    "    sample_list.append(sample)\n",
    "    popnum = linelist[1][-1]\n",
    "    pop_dict[sample] = popnum\n",
    "    \n",
    "lines_for_file = []\n",
    "lines_for_file.append(str_onesnp_file_lines[0])\n",
    "\n",
    "for line in str_onesnp_file_lines[1:]: # exclude header\n",
    "    linelist = line.strip().split()\n",
    "    samplename = linelist[0]\n",
    "    newpopnum = pop_dict[samplename]\n",
    "    genotypes = \"\"\n",
    "    for genotype in linelist[2:]: # exclude sample name and pop id columns\n",
    "        genotypes += genotype + \" \"\n",
    "    genotypes = genotypes[:-1] # get rid of last space\n",
    "    newline = samplename + \"\\t\" + newpopnum + \"\\t\" + genotypes + \"\\n\"\n",
    "    lines_for_file.append(newline)\n",
    "    \n",
    "str_file_w_pops = open(str_filename_w_pops,\"w\")\n",
    "for line in lines_for_file:\n",
    "    str_file_w_pops.write(line)\n",
    "str_file_w_pops.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "deletable": true,
    "editable": true
   },
   "source": [
    "**[9]** Add fake second population to be able to look at populations as one in ``hierfstat``\n",
    "\n",
    "Before I have all of my samples, it makes sense to look at population genetic statistics like Fis distribution and heterozygosity across all of my samples. However, ``hierfstat`` can't calculate these statistics on just one population. So, I can make a false second population just to get it running.\n",
    "\n",
    "Be sure to change the name of the output STR file with false population here:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "outputs": [],
   "source": [
    "str_filename_fpop = \"CG_l1_25_c78_H_biall_NOmaf_oneSNP_inames_fpop.str\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {
    "collapsed": false,
    "deletable": true,
    "editable": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of individuals after adding fpop = 28.0\n",
      "Number of loci = 7272\n"
     ]
    }
   ],
   "source": [
    "str_onesnp_file = open(new_str_filename, \"r\")\n",
    "str_onesnp_file_lines = str_onesnp_file.readlines()\n",
    "str_onesnp_file.close()\n",
    "\n",
    "fpop_lines = []\n",
    "\n",
    "for line in str_onesnp_file_lines[-6:]:\n",
    "    linelist = line.strip().split()\n",
    "    newline = \"FALSE\" + linelist[0] + \"\\t\" + \"2\" + \"\\t\"\n",
    "    for genotype in linelist[2:]:\n",
    "        newline += genotype + \" \"\n",
    "    newline = newline[:-1] + \"\\n\"\n",
    "    fpop_lines.append(newline)\n",
    "\n",
    "str_file_fpop =open(str_filename_fpop, \"w\")\n",
    "\n",
    "for line in str_onesnp_file_lines:\n",
    "    str_file_fpop.write(line)\n",
    "for line in fpop_lines:\n",
    "    str_file_fpop.write(line)\n",
    "str_file_fpop.close()\n",
    "\n",
    "numloci = len(str_onesnp_file_lines[0].strip().split())\n",
    "numinds = int(float(len(fpop_lines)+len(str_onesnp_file_lines)-1))/float(2)\n",
    "\n",
    "print \"Number of individuals after adding fpop = \" + str(numinds) \n",
    "print \"Number of loci = \" + str(numloci)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "collapsed": true,
    "deletable": true,
    "editable": true
   },
   "source": [
    "## Get pop gen stats with heirfstat"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "I made an R script that will read the structure file in with adegenet, and then calculate Ho, Hs, and Fis across loci. Script [here](https://github.com/nclowell/RAD_Scallops/blob/master/CRAGIG_combined_lanes/Scripts/pop_gen_stats.R)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "u'/mnt/hgfs/SHARED_FOLDER/Git_repo/CRAGIG_combined_lanes/Notebooks'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pwd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "# get working directory from results to cell above\n",
    "wd = \"/mnt/hgfs/SHARED_FOLDER/Learn_iPyrad/CRAGIG_RUNS_12_py/CG_l1_25_c78_H_outfiles/new_outfiles\"\n",
    "\n",
    "# 8 digit date for naming files\n",
    "date = \"20170713\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0"
      ]
     },
     "execution_count": 77,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# run R script\n",
    "\n",
    "import subprocess as sp\n",
    "\n",
    "string_callR = \"Rscript ../../../../pop_gen_stats.R \"\n",
    "string_callR += wd + \" \"\n",
    "string_callR += str_filename_fpop + \" \"\n",
    "string_callR += str(numinds) + \" \" + str(numloci) + \" \" + date\n",
    "\n",
    "sp.call([string_callR],shell=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Plots\n",
    "\n",
    "The code here comes from my [Exploratory Plots notebook](https://github.com/nclowell/RAD_Scallops/blob/master/CRAGIG_combined_lanes/Notebooks/Exploratory%20Plots.ipynb). For streamlining my workflow, it's handy to have here too."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\u001b[0m\u001b[01;32mFis_20170711_fpop.txt\u001b[0m*  \u001b[01;32mHo_20170711_fpop.txt\u001b[0m*  \u001b[01;32mHs_20170711_fpop.txt\u001b[0m*\r\n",
      "\u001b[01;32mFis_ 20170713 .txt\u001b[0m*     \u001b[01;32mHo_ 20170713 .txt\u001b[0m*     \u001b[01;32mHs_ 20170713 .txt\u001b[0m*\r\n"
     ]
    }
   ],
   "source": [
    "ls *.txt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "vcf_filename = vcf_filename\n",
    "assembly_name_plots = vcf_filename.split(\".\")[0]\n",
    "str_filename = str_filename_fpop\n",
    "fis_filename = \"Fis_\" + date + \".txt\"\n",
    "ho_filename = \"Ho_\" + date + \".txt\"\n",
    "hs_filename = \"Hs_\" + date + \".txt\"\n",
    "clust_sim = \"78%\" # remove after 20170713"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "vcf_file = open(vcf_filename,\"r\") # change!\n",
    "vcf_file_lines = vcf_file.readlines()\n",
    "vcf_file.close()\n",
    "\n",
    "locus_names = {}\n",
    "snp_name_count = 1\n",
    "\n",
    "for line in vcf_file_lines:\n",
    "    if line[0] != \"#\":\n",
    "        linelist = line.strip().split()\n",
    "        locus = linelist[0]\n",
    "        pos = linelist[1]\n",
    "        vcf_locus_name = locus + \"_\" + pos\n",
    "        str_locus_name = \"SNP_\" + str(snp_name_count)\n",
    "        snp_name_count += 1\n",
    "        locus_names[str_locus_name] = vcf_locus_name  \n",
    "\n",
    "sample_list = []\n",
    "snp_name_list = []\n",
    "rdd = {} \n",
    "\n",
    "for line in vcf_file_lines:\n",
    "    if line[0:1] == \"#C\":\n",
    "        linelist = line.strip().split()\n",
    "        sample_list += linelist[9:]\n",
    "    elif line[0] != \"#\": # ignore header lines\n",
    "        genblocks = line.strip().split()\n",
    "        snp_name = genblocks[0] + \"_\" + genblocks[1] # name formate.g., locus_47_1\n",
    "        snp_name_list.append(snp_name)\n",
    "        for genblock in genblocks[9:]: # start on genotype lines\n",
    "            genblocklist = genblock.split(\":\")\n",
    "            rd = int(genblocklist[1])\n",
    "            if snp_name not in rdd:\n",
    "                rdd[snp_name] = [rd]\n",
    "            elif snp_name in rdd:\n",
    "                rdd[snp_name] += [rd]\n",
    "            else:\n",
    "                print \"something funky going on\"\n",
    "\n",
    "locus_rd_avs = []\n",
    "for snp in snp_name_list:\n",
    "    locus_rd_avs.append(np.mean(rdd[snp]))\n",
    "    \n",
    "plt.hist(locus_rd_avs, bins = np.arange(0,max(locus_rd_avs)+1,4)-2)\n",
    "plt.title(\"Average locus read depth \" + vcf_filename + \"\\n\" + clust_sim)\n",
    "plt.xlabel(\"Read depth\")\n",
    "plt.ylabel(\"Frequency\")\n",
    "\n",
    "plt.savefig('rd_' + assembly_name_plots + '.png')\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {
    "collapsed": false,
    "scrolled": true
   },
   "outputs": [],
   "source": [
    "fis_file = open(fis_filename,\"r\")\n",
    "fis_file_lines = fis_file.readlines()\n",
    "fis_file.close()\n",
    "\n",
    "fis_list = []\n",
    "fis_pls_NAs = []\n",
    "\n",
    "for line in fis_file_lines[1:]:\n",
    "    fis = line.strip().split()[1]\n",
    "    fis_pls_NAs.append(fis)\n",
    "    if fis != \"NA\":\n",
    "        fis_list.append(float(fis))\n",
    "    \n",
    "plt.hist(fis_list, bins = np.arange(-1.2,1.2,.05)-.025)\n",
    "\n",
    "\n",
    "plt.title(fis_filename + \"\\n\" + clust_sim)\n",
    "\n",
    "plt.xlabel(\"Fis\")\n",
    "plt.ylabel(\"Frequency\")\n",
    "\n",
    "plt.savefig(\"Fis_\" + assembly_name_plots + '.png')\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "ho_file = open(ho_filename,\"r\")\n",
    "ho_file_lines = ho_file.readlines()\n",
    "ho_file.close()\n",
    "\n",
    "ho_list = []\n",
    "ho_and_NAs = []\n",
    "\n",
    "for line in ho_file_lines[1:]:\n",
    "    ho = line.strip().split()[1]\n",
    "    ho_and_NAs.append(ho)\n",
    "    if ho != \"NA\":\n",
    "        ho_list.append(float(ho))\n",
    "    \n",
    "plt.hist(ho_list, bins = np.arange(0,1.2,.05)-.025)\n",
    "\n",
    "plt.title(ho_filename + \"\\n\" + clust_sim)\n",
    "plt.xlabel(\"Ho\")\n",
    "plt.ylabel(\"Frequency\")\n",
    "plt.savefig(\"Ho_\" + assembly_name_plots + '.png')\n",
    "plt.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "hs_file = open(hs_filename,\"r\")\n",
    "hs_file_lines = hs_file.readlines()\n",
    "hs_file.close()\n",
    "\n",
    "hs_list = []\n",
    "hs_and_NAs = []\n",
    "\n",
    "for line in hs_file_lines[1:]:\n",
    "    hs = line.strip().split()[1]\n",
    "    hs_and_NAs.append(hs)\n",
    "    if hs != \"NA\":\n",
    "        hs_list.append(float(hs))\n",
    "    \n",
    "plt.hist(hs_list, bins = np.arange(0,1.2,.025)-.0125)\n",
    "\n",
    "plt.title(hs_filename + \"\\n\" + clust_sim)\n",
    "\n",
    "plt.xlabel(\"Hs\")\n",
    "plt.ylabel(\"Frequency\")\n",
    "plt.savefig(\"Hs_\" + assembly_name_plots + '.png')\n",
    "plt.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
