{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Evaluating whether Bowtie and Bowtie2 work differently in filtering stage\n",
    "\n",
    "**20170515**\n",
    "\n",
    "Taken from [the Bowtie2 FAQ web page](http://bowtie-bio.sourceforge.net/bowtie2/faq.shtml), the main differences between Bowtie and Bowtie2 are:\n",
    "* Bowtie 2 fully supports gapped alignment with affine gap penalties. Number of gaps and gap lengths are not restricted, except via the user-supplied scoring scheme. Bowtie 1 only finds ungapped alignments.\n",
    "* For reads longer than about 50 bp Bowtie 2 is generally faster, more sensitive, and uses less memory than Bowtie * For relatively short reads (e.g. less than 50 bp) Bowtie 1 is sometimes faster and/or more sensitive.\n",
    "* Bowtie 2 supports a \"local\" alignment mode, which doesn't require that reads align end-to-end. This produces alignments that might be \"trimmed\" (or \"soft clipped\") at one or both extremes in a way that optimizes alignment score. Bowtie 2 also supports an \"end-to-end\" alignment mode which, like Bowtie 1, requires that the read align entirely.\n",
    "* There is no upper limit on read length in Bowtie 2. Bowtie 1 had an upper limit of around 1000 bp.\n",
    "* Bowtie 2 does away with Bowtie 1's notion of alignment \"stratum\". In Bowtie 2 all alignments lie along a continuous spectrum of alignment scores.\n",
    "* There is no longer a distinction between \"end-to-end\" and \"Maq-like\" modes as in Bowtie 1. There is just one scoring scheme, similar to Needleman-Wunsch and Smith-Waterman.\n",
    "* Bowtie 2's paired-end alignment mode is more flexible than Bowtie 1's. For example, for pairs that do not align in a paired fashion, it will attempt to find unpaired alignments for each mate.\n",
    "* Bowtie 2 does not align colorspace reads.\n",
    "\n",
    "Given that Bowtie2 fully supports gapped alignments, I think it's worth evaluating how much this affects the filtering stage in our pipeline, especially given the amount of genetic variation and indels in my data compared to that of my peers.\n",
    "\n",
    "I verified that the output SAM file formats are about the same, so I'm going to run the Bowtie filtering steps in this notebook using both Bowtie and Bowtie2 to compare results.\n",
    "\n",
    "### Bowtie2\n",
    "\n",
    "I'm going to use the same bowtie index and fasta file of loci to make an alignment with Bowtie2, and compare results to those in the Bowtie filtering step of this [CRAGIG_RUN1 notebook](https://github.com/nclowell/RAD_Scallops/blob/master/CRAGIG_run1/Notebooks/data_processing/CRAGIG_RUN1.ipynb).\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "/mnt/hgfs/SHARED_FOLDER/WorkingFolder/Bowtie/bowtie-1.2\n"
     ]
    }
   ],
   "source": [
    "cd /mnt/hgfs/SHARED_FOLDER/WorkingFolder/Bowtie/bowtie-1.2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(INFO): Before arg handling:\r\n",
      "(INFO):   Wrapper args:\r\n",
      "[  ]\r\n",
      "(INFO):   Binary args:\r\n",
      "[ -f -v 3 --sam --sam-nohead batch_100 seqsforBOWTIE.fa batch_100_BOWTIEout_bowtie2_test.sam ]\r\n",
      "(INFO): After arg handling:\r\n",
      "(INFO):   Binary args:\r\n",
      "[ -f -v 3 --sam --sam-nohead batch_100 seqsforBOWTIE.fa batch_100_BOWTIEout_bowtie2_test.sam ]\r\n",
      "(INFO): Cannot find any index option (--reference-string, --ref-string or -x) in the given command line.\r\n",
      "(INFO): Using the small index (1.1.bt2).\r\n",
      "(INFO): \"/usr/bin/bowtie2-align-s\" --wrapper basic-0 -f -v 3 --sam --sam-nohead batch_100 seqsforBOWTIE.fa batch_100_BOWTIEout_bowtie2_test.sam\r\n",
      "Bowtie 2 version 2.2.6 by Ben Langmead (langmea@cs.jhu.edu, www.cs.jhu.edu/~langmea)\r\n",
      "Usage: \r\n",
      "  bowtie2 [options]* -x <bt2-idx> {-1 <m1> -2 <m2> | -U <r>} [-S <sam>]\r\n",
      "\r\n",
      "  <bt2-idx>  Index filename prefix (minus trailing .X.bt2).\r\n",
      "             NOTE: Bowtie 1 and Bowtie 2 indexes are not compatible.\r\n",
      "  <m1>       Files with #1 mates, paired with files in <m2>.\r\n",
      "             Could be gzip'ed (extension: .gz) or bzip2'ed (extension: .bz2).\r\n",
      "  <m2>       Files with #2 mates, paired with files in <m1>.\r\n",
      "             Could be gzip'ed (extension: .gz) or bzip2'ed (extension: .bz2).\r\n",
      "  <r>        Files with unpaired reads.\r\n",
      "             Could be gzip'ed (extension: .gz) or bzip2'ed (extension: .bz2).\r\n",
      "  <sam>      File for SAM output (default: stdout)\r\n",
      "\r\n",
      "  <m1>, <m2>, <r> can be comma-separated lists (no whitespace) and can be\r\n",
      "  specified many times.  E.g. '-U file1.fq,file2.fq -U file3.fq'.\r\n",
      "\r\n",
      "Options (defaults in parentheses):\r\n",
      "\r\n",
      " Input:\r\n",
      "  -q                 query input files are FASTQ .fq/.fastq (default)\r\n",
      "  --qseq             query input files are in Illumina's qseq format\r\n",
      "  -f                 query input files are (multi-)FASTA .fa/.mfa\r\n",
      "  -r                 query input files are raw one-sequence-per-line\r\n",
      "  -c                 <m1>, <m2>, <r> are sequences themselves, not files\r\n",
      "  -s/--skip <int>    skip the first <int> reads/pairs in the input (none)\r\n",
      "  -u/--upto <int>    stop after first <int> reads/pairs (no limit)\r\n",
      "  -5/--trim5 <int>   trim <int> bases from 5'/left end of reads (0)\r\n",
      "  -3/--trim3 <int>   trim <int> bases from 3'/right end of reads (0)\r\n",
      "  --phred33          qualities are Phred+33 (default)\r\n",
      "  --phred64          qualities are Phred+64\r\n",
      "  --int-quals        qualities encoded as space-delimited integers\r\n",
      "\r\n",
      " Presets:                 Same as:\r\n",
      "  For --end-to-end:\r\n",
      "   --very-fast            -D 5 -R 1 -N 0 -L 22 -i S,0,2.50\r\n",
      "   --fast                 -D 10 -R 2 -N 0 -L 22 -i S,0,2.50\r\n",
      "   --sensitive            -D 15 -R 2 -N 0 -L 22 -i S,1,1.15 (default)\r\n",
      "   --very-sensitive       -D 20 -R 3 -N 0 -L 20 -i S,1,0.50\r\n",
      "\r\n",
      "  For --local:\r\n",
      "   --very-fast-local      -D 5 -R 1 -N 0 -L 25 -i S,1,2.00\r\n",
      "   --fast-local           -D 10 -R 2 -N 0 -L 22 -i S,1,1.75\r\n",
      "   --sensitive-local      -D 15 -R 2 -N 0 -L 20 -i S,1,0.75 (default)\r\n",
      "   --very-sensitive-local -D 20 -R 3 -N 0 -L 20 -i S,1,0.50\r\n",
      "\r\n",
      " Alignment:\r\n",
      "  -N <int>           max # mismatches in seed alignment; can be 0 or 1 (0)\r\n",
      "  -L <int>           length of seed substrings; must be >3, <32 (22)\r\n",
      "  -i <func>          interval between seed substrings w/r/t read len (S,1,1.15)\r\n",
      "  --n-ceil <func>    func for max # non-A/C/G/Ts permitted in aln (L,0,0.15)\r\n",
      "  --dpad <int>       include <int> extra ref chars on sides of DP table (15)\r\n",
      "  --gbar <int>       disallow gaps within <int> nucs of read extremes (4)\r\n",
      "  --ignore-quals     treat all quality values as 30 on Phred scale (off)\r\n",
      "  --nofw             do not align forward (original) version of read (off)\r\n",
      "  --norc             do not align reverse-complement version of read (off)\r\n",
      "  --no-1mm-upfront   do not allow 1 mismatch alignments before attempting to\r\n",
      "                     scan for the optimal seeded alignments\r\n",
      "  --end-to-end       entire read must align; no clipping (on)\r\n",
      "   OR\r\n",
      "  --local            local alignment; ends might be soft clipped (off)\r\n",
      "\r\n",
      " Scoring:\r\n",
      "  --ma <int>         match bonus (0 for --end-to-end, 2 for --local) \r\n",
      "  --mp <int>         max penalty for mismatch; lower qual = lower penalty (6)\r\n",
      "  --np <int>         penalty for non-A/C/G/Ts in read/ref (1)\r\n",
      "  --rdg <int>,<int>  read gap open, extend penalties (5,3)\r\n",
      "  --rfg <int>,<int>  reference gap open, extend penalties (5,3)\r\n",
      "  --score-min <func> min acceptable alignment score w/r/t read length\r\n",
      "                     (G,20,8 for local, L,-0.6,-0.6 for end-to-end)\r\n",
      "\r\n",
      " Reporting:\r\n",
      "  (default)          look for multiple alignments, report best, with MAPQ\r\n",
      "   OR\r\n",
      "  -k <int>           report up to <int> alns per read; MAPQ not meaningful\r\n",
      "   OR\r\n",
      "  -a/--all           report all alignments; very slow, MAPQ not meaningful\r\n",
      "\r\n",
      " Effort:\r\n",
      "  -D <int>           give up extending after <int> failed extends in a row (15)\r\n",
      "  -R <int>           for reads w/ repetitive seeds, try <int> sets of seeds (2)\r\n",
      "\r\n",
      " Paired-end:\r\n",
      "  -I/--minins <int>  minimum fragment length (0)\r\n",
      "  -X/--maxins <int>  maximum fragment length (500)\r\n",
      "  --fr/--rf/--ff     -1, -2 mates align fw/rev, rev/fw, fw/fw (--fr)\r\n",
      "  --no-mixed         suppress unpaired alignments for paired reads\r\n",
      "  --no-discordant    suppress discordant alignments for paired reads\r\n",
      "  --no-dovetail      not concordant when mates extend past each other\r\n",
      "  --no-contain       not concordant when one mate alignment contains other\r\n",
      "  --no-overlap       not concordant when mates overlap at all\r\n",
      "\r\n",
      " Output:\r\n",
      "  -t/--time          print wall-clock time taken by search phases\r\n",
      "  --un <path>           write unpaired reads that didn't align to <path>\r\n",
      "  --al <path>           write unpaired reads that aligned at least once to <path>\r\n",
      "  --un-conc <path>      write pairs that didn't align concordantly to <path>\r\n",
      "  --al-conc <path>      write pairs that aligned concordantly at least once to <path>\r\n",
      "  (Note: for --un, --al, --un-conc, or --al-conc, add '-gz' to the option name, e.g.\r\n",
      "  --un-gz <path>, to gzip compress output, or add '-bz2' to bzip2 compress output.)\r\n",
      "  --quiet            print nothing to stderr except serious errors\r\n",
      "  --met-file <path>  send metrics to file at <path> (off)\r\n",
      "  --met-stderr       send metrics to stderr (off)\r\n",
      "  --met <int>        report internal counters & metrics every <int> secs (1)\r\n",
      "  --no-unal          supppress SAM records for unaligned reads\r\n",
      "  --no-head          supppress header lines, i.e. lines starting with @\r\n",
      "  --no-sq            supppress @SQ header lines\r\n",
      "  --rg-id <text>     set read group id, reflected in @RG line and RG:Z: opt field\r\n",
      "  --rg <text>        add <text> (\"lab:value\") to @RG line of SAM header.\r\n",
      "                     Note: @RG line only printed when --rg-id is set.\r\n",
      "  --omit-sec-seq     put '*' in SEQ and QUAL fields for secondary alignments.\r\n",
      "\r\n",
      " Performance:\r\n",
      "  -p/--threads <int> number of alignment threads to launch (1)\r\n",
      "  --reorder          force SAM output order to match order of input reads\r\n",
      "  --mm               use memory-mapped I/O for index; many 'bowtie's can share\r\n",
      "\r\n",
      " Other:\r\n",
      "  --qc-filter        filter out reads that are bad according to QSEQ filter\r\n",
      "  --seed <int>       seed for random number generator (0)\r\n",
      "  --non-deterministic seed rand. gen. arbitrarily instead of using read attributes\r\n",
      "  --version          print version information and quit\r\n",
      "  -h/--help          print this usage message\r\n",
      "Error: Encountered internal Bowtie 2 exception (#1)\r\n",
      "Command: /usr/bin/bowtie2-align-s --wrapper basic-0 -f -v 3 --sam --sam-nohead batch_100 seqsforBOWTIE.fa batch_100_BOWTIEout_bowtie2_test.sam \r\n",
      "(ERR): bowtie2-align exited with value 1\r\n"
     ]
    }
   ],
   "source": [
    "# f refers to input file format as fasta\n",
    "# v refers to the number of mismatches allowed\n",
    "# --sam refers to the option to print alignments in SAM format\n",
    "# --sam-nohead to repress header\n",
    "!bowtie2 -f -v 3 --sam --sam-nohead \\\n",
    "batch_100 \\\n",
    "seqsforBOWTIE.fa \\\n",
    "batch_100_BOWTIEout_bowtie2_test.sam"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
